### ISSUE TITLE
Lazy sequence concatenation in C# and Java runtimes is not thread-safe

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#1454]
Both [C#](https://github.com/dafny-lang/dafny/blob/90eb1084d9862384accd03fb5a1c8af500212a05/Source/DafnyRuntime/DafnyRuntime.cs#L1111-L1120) and [Java](https://github.com/dafny-lang/dafny/blob/90eb1084d9862384accd03fb5a1c8af500212a05/Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java#L756-L761) optimize sequence concatenation by creating a lazy representation with internal mutable state. I've suspected this isn't thread safe and have now confirmed it isn't in C# at least, using Coyote. Namely, two threads can both determine that the lazy representation needs to be resolved, one thread will complete this operation and clear the `left` and `right` fields so they can be GC'd, and then the second thread throws a NullReferenceException when trying to do the same.

One possible answer to this is "not an issue, the runtimes and Dafny itself aren't guaranteed to be thread-safe", but I'd be personally disappointed by this given this seems to be the only such issue with the runtimes AFAICT so far. The solution in C# is likely to use the fairly sophisticated and thread-safe `ImmutableList` class instead of our hand-rolled implementation of this immutable datatype, and the Java runtime likely has something similar we can wrap.

See also https://github.com/dafny-lang/rfcs/pull/6.

**Comment by robin-aws** [comment]
For whoever picks this up: the Coyote test I created to confirm this bug in C# is here, from my fork: https://github.com/robin-aws/dafny/blob/thread-safe-runtimes/Source/DafnyRuntime.Test/UnitTest1.cs

The acceptance criteria for this fix should be to add tests like that to CI, so we have a mechanism to prevent regressions.

**Comment by keyboardDrummer** [comment]
@robin-aws could you sketch the setup of an application that needs Dafny sequences to be thread-safe? Is some host application calling Dafny code, getting Dafny sequences from them, and then accessing them using multiple threads?

Would it not be possible to provide the thread-safety from the host context, by synchronising around the sequence access?

**Comment by robin-aws** [comment]
Reopening as this is was fixed for C# in #1780 but is still an issue for Java.

@keyboardDrummer, apologies for not spotting your comment early. Here's a super contrived example: 

```dafny
module NotThreadSafe {
  const SomeString := "Hello " + "World!";
  function method {:extern} GetFirstLetter(): char {
    SomeString[0]
  }
}
```

If multiple threads call `GetFirstLetter` simultaneously, they will both trigger the lazy evaluation of the concatenation, and could potentially return the wrong result or crash.

Since the sources of such thread un-safety are not tracked by the language and can be arbitrarily deep, the only safe solution is to use a "global interpreter lock" around the execution of any Dafny code. That's a very expensive solution and disappointing when your Dafny code otherwise uses only immutable data and shouldn't have to worry about this issue.
